### PR TITLE
Reduce code duplication between DmaFile and Directory

### PR DESCRIPTION
**Pull request by matklad**
### What does this PR do?

Cleanup discussed in #102 

### Checklist

[] I have added unit tests to the code I am submitting
[] My unit tests cover both failure and success scenarios
[] If applicable, I have discussed my architecture
[x] The new code I am adding is formatted using `rustfmt`

### CODE REVIEW

File: scipio/src/io/dma_file.rs [hunk 1]
```diff
@@ -462,6 +434,79 @@ impl DmaFile {
     }
 }
 
+/// A wrapper over `std::fs::File` which carries a path (for better error
+/// messages) and prints a warning if closed synchronously.
+#[derive(Debug)]
+struct EnhancedFile {
```
Comment (matklad) [inline scipio/src/io/dma_file.rs:440]: I am not thrilled about the name tbh. I *think* I would like `ScipioFile` more, but it seems that we generally don't use "Scipio" prefix anywhere else?

Comment (glommer) [reply scipio/src/io/dma_file.rs:440]: `ScipioFile` is fine as long as it doesn't get confusing.

`Directory` is well, a directory.
`DmaFile` allows for dma operations (O_DIRECT) randomly issued at direct positions.
`StreamReader` and `StreamWriter` allow for linear access (currently only over O_DIRECT).
I have a queued patchset for `BufferedFile` which will use buffered I/O for the simple applications that don't want to deal with O_DIRECT.

So I wouldn't be opposed to `ScipioFile` as something common to *all* of those.

Closing seems like something that will be shared between `Directory`, `BufferedFile` and `DmaFile`. It's fair to say that the streams are not *really* files, but rather abstractions on top of file.

So maybe?

**Comment by glommer** [comment]
Good change and I could merge it as is.

Please consider folding your last patch into the first so it is called `ScipioFile` from the beginning.
If you post a fast forward with an empty diff from b9c25f6 I'll merge

**Comment by matklad** [comment]
Rebased! (In general, I am very pro-rebase and skilled with magit, so feel free to request history cleanups)

**Comment by glommer** [comment]
Shh, don't let the linux folks hear it or you'd end up doing kernel work

**Pull request merged**